Library/Homebrew/dependency_collector.rb

```diff
@@ -28,6 +28,14 @@ class DependencyCollector
   def initialize
     @deps = Dependencies.new
     @requirements = Requirements.new
+
+    init_global_dep_tree
+  end
+
+  def initialize_copy(other)
```

Comment:
SoftwareSpec's are created on the class-level of a formula rather than instance level. We create a copy of this in `spec_eval` when creating a formula instance.

This copy is however a shallow copy. It does not copy the dependency collector so when we call `add_global_deps_to_spec` we are actually affecting the original version of the SoftwareSpec rather than just the copy. This leads to multiple copies of the global deps being added to the collector every time the formula instance is created.

The adjustment here is `initialize_copy` (which is called by Ruby on `.dup` and `.clone`) to `.dup` the dependency collector, and again within the collector to `.dup` the dependency and requirement lists.

Reply:
> The adjustment here is implementing initialize_copy (which is called by Ruby on .dup and .clone) in SoftwareSpec to .dup the dependency collector, and again within the collector to .dup the dependency and requirement lists.

TIL, thanks!